data/exploits/CVE-2022-22942/cve-2022-22942-dc.c

```diff
@@ -0,0 +1,601 @@
+/* The vmwgfx driver has a similar bug as the one we fixed last year in the
+ * nitro enclaves code (https://git.kernel.org/linus/f1ce3986baa6
+ * "nitro_enclaves: Fix stale file descriptors on failed usercopy").
+ *
+ * If the driver fails to copy the 'fence_rep' object to userland, it tries to
+ * recover by deallocating the (already populated) file descriptor. This is
+ * wrong, as the fd gets released via put_unused_fd() which shouldn't be used,
+ * as the fd table slot was already populated via the previous call to
+ * fd_install(). This leaves userland with a valid fd table entry pointing to
+ * a free'd 'file' object.
+ *
+ * There are multiple ways to exploit this bug. A previous version of this PoC
+ * dumped the contents of /etc/shadow. This one overwrites a SUID root binary
+ * to pop a shell.
+ *
+ * Compile as:
+ *   $ gcc -O2 cve-2022-22942-dc.c -o cve-2022-22942-dc
+ *
+ * Run as (and wait for the root shell to appear):
+ *   $ ./cve-2022-22942-dc [target_file [temp_file [dev_node]]]
+ *
+ * Remarks:
+ *
+ * This POC assumes it has access to '/dev/dri/card0' which likely means the
+ * calling user needs to be part of the 'video' group.
+ *
+ * Alternatively '/dev/dri/renderD128' can be used (just pass the path as
+ * argument to ./cve-2022-22942-dc-dc), which, under Debian, means being part
+ * of the 'render' group.
+ *
+ * This bug was fixed by commit a0f90c881570 ("drm/vmwgfx: Fix stale file
+ * descriptors on failed usercopy"). It affected kernel versions v4.14-rc1 to
+ * v5.17-rc1.
+ *
+ * This is CVE-2022-22942.
+ *
+ * (c) 2022 Open Source Security, Inc.
+ *
+ * - minipli
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
```

Comment:
Are we allowed to accept this license? 👀

Reply:
Hey @h00die, it's come to my attention that we actually can't really accept code with a GNU General Public License. I would hate for the effort put into this module (and https://github.com/rapid7/metasploit-framework/pull/17301) to go to waste. 

Would you maybe be open to asking the original author if they would consider changing the license to one that we accept? If not we'd have to re-implement the C code which would be a pain. 

=========== Acceptable Licenses at Rapid7 ===============
Third Party Software dedicated to the public domain or licensed
pursuant to permissive licenses which do not restrict use,
modification, distribution, and hosting of the Open Source Software
(such as [Apache v1.1](https://opensource.org/licenses/Apache-1.1), [Apache v2.0](https://opensource.org/licenses/Apache-2.0),[ BSD-style](https://opensource.org/licenses/BSD-3-Clause), [MIT-style](https://opensource.org/licenses/MIT), [PostgreSQL](https://opensource.org/licenses/PostgreSQL),
[Boost](https://opensource.org/licenses/BSL-1.0), [Zlib/libpng](https://opensource.org/licenses/Zlib), and [W3C](https://opensource.org/licenses/W3C) licenses).

**Unmodified** Third Party Software licensed pursuant to Public
Licenses (such as [Mozilla Public License (MPL) v2.0](https://opensource.org/licenses/MPL-2.0), [Common
Development and Distribution License (CDDL)](https://opensource.org/licenses/CDDL-1.0), [Eclipse Public License
(EPL)](https://opensource.org/licenses/EPL-1.0), [Common Public License](https://opensource.org/licenses/CPL-1.0), and [Microsoft Public
License (Ms-PL)](https://opensource.org/licenses/MS-PL).

My apologies for just bringing this up now, I was unaware of this policy when I started the review. Really hoping we can land these two modules 🤞.